crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -483,54 +482,74 @@ impl OlmMachine {
-    /// sent, the responses should be passed back to the state machine using
-    /// [`mark_request_as_sent`].
+    /// A triple of requests which should be sent out to the server, in the
+    /// order they appear in the return tuple.
+    ///
+    /// The second request's response, if present, should be passed back to the
+    /// state machine using [`mark_request_as_sent`].
     ///
     /// These requests may require user interactive auth.
     ///
     /// [`mark_request_as_sent`]: #method.mark_request_as_sent
     pub async fn bootstrap_cross_signing(
         &self,
         reset: bool,
-    ) -> StoreResult<(UploadSigningKeysRequest, UploadSignaturesRequest)> {
+    ) -> StoreResult<(UploadSigningKeysRequest, Option<OutgoingRequest>, UploadSignaturesRequest)>
+    {
         let mut identity = self.inner.user_identity.lock().await;
 
         if reset || identity.is_empty().await {
             info!("Creating new cross signing identity");
             let cache = self.inner.store.cache().await?;
             let account = cache.account().await?;
-            let (id, request, signature_request) = account.bootstrap_cross_signing().await;
 
-            *identity = id;
+            let (new_identity, upload_signing_keys_req, upload_signatures_req) =
+                account.bootstrap_cross_signing().await;
+
+            // If there are any keys to upload, upload them before we upload the signatures,
+            // since the signatures may reference keys to be uploaded.
+            let upload_keys_req = self.keys_for_upload(&account).await.map(OutgoingRequest::from);
+
+            *identity = new_identity;
 
             let public = identity.to_public_identity().await.expect(
                 "Couldn't create a public version of the identity from a new private identity",
             );
 
-            let changes = Changes {
-                identities: IdentityChanges { new: vec![public.into()], ..Default::default() },
-                private_identity: Some(identity.clone()),
-                ..Default::default()
-            };
-
-            self.store().save_changes(changes).await?;
+            self.store()
+                .save_changes(Changes {
+                    identities: IdentityChanges { new: vec![public.into()], ..Default::default() },
+                    private_identity: Some(identity.clone()),
+                    ..Default::default()
+                })
+                .await?;
 
-            Ok((request, signature_request))
+            Ok((upload_signing_keys_req, upload_keys_req, upload_signatures_req))
         } else {
             info!("Trying to upload the existing cross signing identity");
-            let request = identity.as_upload_request().await;
-            let account = self.inner.store.static_account();
+            let upload_signing_keys_req = identity.as_upload_request().await;
+
+            // If there are any keys to upload, upload them before we upload the signatures,
```

Comment:
I think we should only do this if the `shared` flag of the `Account` wasn't set to `true`. `keys_for_upload` will give you a request any time there are some one-time keys to be uploaded, but we only need the device keys to be published which the `shared` flag remembers.